LSPS1/extensions/lnurl_lud2/LSPS1.lnurl_lud2.yaml

```diff
@@ -0,0 +1,58 @@
+openapi: 3.0.3
+info:
+  description: |
+    API documentation for channel opening request from a Lightning Network Service Provider (LSP). LNURL_LUD2.
```

Comment:
I am trying to answer all your questions (also continuing from #18). Happy to be convinced to remove this extension.

### LNURL support
I don't see how we can support LNURL in general if we don't provide any additional endpoint. [LUD-02](https://github.com/lnurl/luds/blob/luds/02.md) requires a very specific callback url (<callback><?/&>k1=<k1>&remoteid=<Local LN node ID>&private=<1/0>) that needs to be supported to open the channel. Either we provide the LNURL string, or provide the required callback url with a standardized way of calling the api.

### Channel open on peer connect
I dislike the idea of opening the channel as soon as the peer connects. There is no proper request/response way to return an error code in case the channel open fails.

Let me know what you think.

Reply:
As I mentioned in the meeting as well, if the channel order is paid via Lightning, it is possible to send back the error message over Lightning as well, and it is also an automatic refund mechanism.

For onchain, you need a separate mechanism to handle refunds anyway (LNURL does not provide this to you anyway), maybe that should be the failed-to-open-channel-error-reporting access point.

I do not really give much credence to manual flows where some web app accesses this API and then is a very thin UI layer on top of this API for initial users that need to use existing LNURL mechanisms.  This is supposed to be an *Application Program* Interface, not a manually-operated *User* Interface.

Reply:
I believe the confusion we are having in the discussion stems from the fact that we have different requirements. Please have a look at the [motivation section](https://github.com/synonymdev/lsp-spec/blob/channel-request2/LSPS1/README.md#motivation) that I added with my recent commits.

> I just prefer the simplest approach because it means less code and less room for mistakes.

I am ready to discuss that our approach is too complex. I believe the requirement of "bitcoin onchain payments" (which is necessary to onboard new users to Lightning and to push sats to the user) makes the whole api more complicated. Maybe this can be solved and therefore removed with dual funded channels?

> As I mentioned in the meeting as well, if the channel order is paid via Lightning, it is possible to send back the error message over Lightning as well, and it is also an automatic refund mechanism.

I am happy to discuss if we should also send back the error message via Lightning.

> For onchain, you need a separate mechanism to handle refunds anyway (LNURL does not provide this to you anyway), maybe that should be the failed-to-open-channel-error-reporting access point.

Yes, it's on my todo

> I do not really give much credence to manual flows where some web app accesses this API and then is a very thin UI layer on top of this API for initial users that need to use existing LNURL mechanisms. This is supposed to be an Application Program Interface, not a manually-operated User Interface.

See the motivation section. Not sure what you mean with API vs UI.

Reply:
> I am ready to discuss that our approach is too complex. I believe the requirement of "bitcoin onchain payments" (which is necessary to onboard new users to Lightning and to push sats to the user) makes the whole api more complicated. Maybe this can be solved and therefore removed with dual funded channels?

Indeed, that is the design goal of the dual-funded channel scheme; money is not even held and then refunded in the case where the server cannot (or refuses, for other reasons) serve the request.  It is the reason why we want to consider a simpler API here, as we view this as a stopgap until dual-funding / openV2 is widely deployed later.

Until then, what we need is some kind of refund negotiation for the onchain case.  I would still prefer having to give the refund address up-front, e.g. `refund_onchain_address`.  Need to specify exactly which standards servers MUST support (bech32m, bech32, base58).  For a manually-operated UI you could have the web app hold its own keys in client-side storage and provide the refund address from there, and only if the server refunds, ask the user for a refund address and then move from the web app refund address to the user refund address.  If the web app is run by the same people as the ones running the server it is the same trust requirement anyway, and if you expect the refund case to be rare then the inefficiency of -> web app refund -> user refund is not that important compared to the API complexity (everyone has to implement the API, not everyone has to write a manually-operatable web app, and we expect the refund case to be rare).

Reply:
Did you have a chance to read the two different cases I described sync vs async?
We can use the same strategy for the `refund_onchain_address` as we use with the node_id. You can either provide it upfront (async case) or with the `/lsp/channels/{id}/update` endpoint at a later point (sync case). This would not increase the complexity of the API. The main complexity comes from having local_balance/onchain payments at all.

Reply:
There is no motivation for why you want to provide the sync case.  Additional endpoints = more complexity.  For the webapp case, I do not see why sync case is somehow better than what I described.

Reply:
> One thing we could do is add a "hey, poll now" endpoint, which is way simpler than LNURL-02 but would still accomplish removing the polling complexity from the server end.

Or just make the "LUD-02 endpoint" that endpoint, LOL.  Except LUD-02 expects that it is the endpoint that receives the client node ID, sigh.

Reply:
How do you want to handle the case of a public channel marketplace like for example on Amboss? Requiring the user to add all the data at the beginning (node_id, channel sizes, refund address) will end in a UI/UX disaster. Marketplaces have a very limited amount of data not even talking about the ability to auto-generate addresses.
Additionally, LNURL requires a sync channel open.

Reply:
What information does the marketplace need, anyway?  Just have a separate API for marketplace pricing information.  Have an API do one thing and do it well, not try to throw everything you think of into it, it just becomes a maintenance burden.  Focus this API on buying channels, make a separate API for marketplace info.

Reply:
If you make a marketplace that pings this address... like, WTF, you would effectively DoS the LSPs everytime you queried market pricing information, because each time an LSP is contacted here you put another entry into their database, which the marketplace then does not ever claim.  Like come on.  Do the sensible thing and create a separate `GET`-able endpoint for pricing information that you put in your marketplace.  Could even reuse the caching information of HTTP for how long the pricing information would be valid,

Reply:
I am not intending to push pricing information in to the `POST` request. The point I want to make is:

When a user wants to buy a channel, most (if not all) channel purchase websites use the following UX:
1. The user enters channel size, local balance.
2. The user pays the invoice.
3. The user either uses the LNURL to open the channel OR enters their pubkey manually.
4.1. If success everything is good.
4.2 If fail the user enters the `refund_onchain_address` in case onchain bitcoin was involved. Lightning refunds are done automatically.

Additionally, I would like to include this in the spec (happy to discuss if necessary):
4.3 If fail, the user may correct the `node_connection_string`. This is a major source of errors in case the user enters the node details manually.

The focus of this api is still purchasing channels. Maybe it is better to use the word "websites" and not "marketplace" to avoid unnecessary confusion.

The goal is to provide the lowest UI/UX friction as possible. As John said on the call 2 weeks ago: Yes, we could separate the API in a wallet channel purchase API and a website channel purchase API. The two apis overlap so much though that we would basically build the same thing two times.

Reply:
The website case can be implemented by an LSP even without this API being exposed.  The entire point of this API specification is to have something that wallets can do directly without user intervention.  Which is the whole argument we have been having here: why are you infecting this particular API specifcation with UI concerns of a website?  Implement the two APIs and specify only the one for wallets, because nobody has the wallet use-case covered but everyone can easily build a website even without this spec.  The wallet use-case **needs** this spec.  The webstie use-case can use an internal API that is unspecified to the public.

Reply:
I had a call with John yesterday and we will drop the website case including LNURL for this spec . It takes too much time to get to a common denominator. I tried to incorporate all our previous thoughts/requirements in the spec group up to when Spiral joined. We agree that it is best to focus on the wallet use case, which we know we can use within our own wallet too.

Reply:
I think this makes a lot of sense, afaict, it's great that synonym has a manual flow via a website that essentially lets you "sell an LNURL link." I don't think anyone is questioning that's a valid use case.

Issue for me is that can be incorporated in multiple ways, and if you're expecting someone to come to your website with a wallet that uses LNURL and a method of payment, they don't need to incorporate an API. They already have everything they need and you've done the work to make it easy, which is great.

The real deep purpose for this API for me is so that someone can programmatically request inbound from an LSP, in which case, since they're already doing the work to incorporate this API, we can make the flow streamlined and intuitive, without an LNURL requirement. I see this API as a way for breez to talk to synonym, our LSP to work with X whatever wallet, etc. It'll be easier to get to something we want with a narrower scope at first.

Reply:
I definitely emphasize with your POV and there are no bad feelings in focusing the scope.

The reason that I liked to incorporate LNURL/manual flow was the increased adoption potential. 
A developer friendly API would have increased the likelihood of adoption by a marketplace like Amboss or LN+. LNURL is widely supported in multiple wallets; so incentives would have been there to make more out of this api.

Happy to be called a dreamer in case you disagree with this approach :) But again, I got no issues with a narrower scope.

Reply:
>A developer friendly API would have increased the likelihood of adoption by a marketplace like Amboss or LN+. LNURL is widely supported in multiple wallets; so incentives would have been there to make more out of this api.

I would suggest making a different API, preferably with input from Amboss or LN+ developers, so that you know what they actually want from that API.  Basically we know what *we* want from *this* API, so that is something we can give you input on, and you agreed that the reduced scope is something you could also use for your own plans as well.

For those I imagine the most important information is "how much would it cost to buy inbound liquidity from this LSP" and that is probably best as a separate `GET`-able API.